source/unstructured_source.go

```diff
@@ -0,0 +1,295 @@
+		apiVersion:       apiVersion,
+		kind:             kind,
+		targetPropPath:   strings.Split(targetPropPath, "."),
+		namespace:        namespace,
+		annotationFilter: annotationFilter,
+		labelSelector:    labelSelector,
+		client:           client,
+		codec:            runtime.NewParameterCodec(scheme),
+	}, nil
+}
+
+func (us *unstructuredSource) AddEventHandler(ctx context.Context, handler func()) {
+}
+
+// Endpoints returns endpoint objects.
+func (us *unstructuredSource) Endpoints(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	var endpoints []*endpoint.Endpoint
+
+	result, err := us.List(ctx, &metav1.ListOptions{
+		LabelSelector: us.labelSelector.String(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	result, err = us.filterByAnnotations(result)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range result.Items {
+		item := &result.Items[i]
+
+		hostNames := getHostnamesFromAnnotations(item.GetAnnotations())
+		if len(hostNames) == 0 {
+			log.Warnf("Endpoint %s has no host names", item.GetName())
+			continue
+		}
+
+		var (
+			target      string
+			parentField = item.Object
+		)
+		for j, fieldName := range us.targetPropPath {
+			if j == len(us.targetPropPath)-1 {
+				target, _, _ = unstructured.NestedString(parentField, fieldName)
+			} else {
+				parentField, _, _ = unstructured.NestedMap(parentField, fieldName)
+				if len(parentField) == 0 {
+					break
+				}
+			}
+		}
+		if target == "" {
+			log.Warnf(
+				"Endpoint %s with host names %s has no target",
+				item.GetName(), hostNames)
+			continue
+		}
+
+		ttl, err := getTTLFromAnnotations(item.GetAnnotations())
+		if err != nil {
+			log.Warnf(
+				"Failed to parse TTL for Endpoint %s with host names %s: %s",
+				item.GetName(), hostNames, err)
+			continue
+		}
+
+		// Get the record type, defaulting to A if not found.
+		recordType := getRecordTypeFromAnnotations(item.GetAnnotations())
+		if recordType == "" {
+			recordType = "A"
+		}
+
+		// Create an endpoint for each host name.
+		for _, dnsName := range hostNames {
+			endpoints = append(
+				endpoints,
+				&endpoint.Endpoint{
+					Labels:     map[string]string{},
+					DNSName:    dnsName,
+					Targets:    endpoint.NewTargets(target),
+					RecordType: recordType,
+					RecordTTL:  ttl,
+				},
+			)
+		}
+
+		us.setResourceLabel(item, endpoints)
+	}
+
+	return endpoints, nil
+}
+
+func (us *unstructuredSource) setResourceLabel(
+	obj *unstructured.Unstructured,
+	endpoints []*endpoint.Endpoint) {
+	for _, ep := range endpoints {
+		ep.Labels[endpoint.ResourceLabelKey] = fmt.Sprintf(
+			"unstructured/%s/%s",
+			obj.GetNamespace(),
+			obj.GetName())
+	}
+}
+
+func (us *unstructuredSource) List(
+	ctx context.Context,
+	opts *metav1.ListOptions) (result *unstructured.UnstructuredList, err error) {
+	result = &unstructured.UnstructuredList{}
+	result.SetAPIVersion(us.apiVersion)
+	result.SetKind(us.kind + "List")
+	err = us.client.Get().
+		Namespace(us.namespace).
```

Comment:
Out of curiosity, does this assume that the resource kind specified is a namespaced resource?  If that's the case, do we need to call that somewhere (and make it explicit)?

Reply:
Yeah, I was also looking at that...

Reply:
Okay, it's resolved. The source now supports both cluster _and_ namespace scoped APIs. The API definition on the server determines the scope.

Reply:
@aruneshpa, PTAL at the documentation again -- it now includes an example of a cluster resource as well to highlight that both namespace-scoped and cluster-scoped APIs are supported.